File: python/gameoflife.py
#! /usr/bin/env python3.6
import random
import copy

class Grid:
    grid = []
    width = 0
    height = 0

    def __init__(self, width, height):
        grid = []
        for i in range(0, height):
            grid.append([])
            for _ in range(0, width):
                grid[i].append(0)
        self.grid = grid
        self.width = width
        self.height = height

    def __getitem__(self, key):
        return self.grid[key]

    def __iter__(self):
        return iter(self.grid)

    def randomize(self):
        for row in self.grid:
            for i in range(0, len(row)):
                if random.random() < 0.2:
                    row[i] = 1

class Simulation:
    startGrid = None
    nextState = None
    currentState = None
    lastState = None

    def __init__(self, width=20, height=20, randomize=True):
        self.nextState = Grid(width, height)
        self.currentState = Grid(width, height)
        self.lastState = Grid(width, height)
        if randomize:
            self.randomize_start()

    def randomize_start(self):
        self.currentState.randomize()

    def run(self, limit=0):
        self.startGrid = copy.deepcopy(self.currentState.grid)

        generation = 1

        stalled = False
        while not stalled:
            if limit > 0 and generation == limit:
                break

            stalled = self._next_generation()
            self._rotate_grids()
            generation += 1

        return generation


    def _next_generation(self):
        stalled = True
        for i, row in enumerate(self.currentState):
            for j in range(0, len(row)):
                neighbors = self._count_neighbors(i, j)

                if neighbors < 2:
                    self.nextState[i][j] = 0
                elif (neighbors == 2 or neighbors == 3) and row[j] == 1:
                    self.nextState[i][j] = 1
                elif neighbors > 3:
                    self.nextState[i][j] = 0
                elif neighbors == 3 and row[j] == 0:
                    self.nextState[i][j] = 1
                else:
                    self.nextState[i][j] = 0

                if self.nextState[i][j] != self.lastState[i][j]:
                    stalled = False

        return stalled

    def _rotate_grids(self):
        last = self.lastState
        self.lastState = self.currentState
        self.currentState = self.nextState
        self.nextState = last

    def _count_neighbors(self, row, column):
        count = 0

        if row-1 >= 0:
            if self.currentState[row-1][column] == 1:
                count += 1

        if row+1 < self.currentState.width:
            if self.currentState[row+1][column] == 1:
                count += 1

        if column-1 >= 0:
            if self.currentState[row][column-1] == 1:
                count += 1

        if column+1 < self.currentState.height:
            if self.currentState[row][column+1] == 1:
                count += 1

        if column-1 >= 0 and row-1 >= 0:
            if self.currentState[row-1][column-1] == 1:
                count += 1

        if column-1 >= 0 and row+1 < self.currentState.width:
            if self.currentState[row+1][column-1] == 1:
                count += 1

        if column+1 < self.currentState.height and row-1 >= 0:
            if self.currentState[row-1][column+1] == 1:
                count += 1

        if column+1 < self.currentState.height and row+1 < self.currentState.width:
            if self.currentState[row+1][column+1] == 1:
                count += 1

        return count

def main():
    sim = Simulation()
    print("Simulation stalled in {} generations".format(sim.run()))


if __name__ == '__main__':
    main()

File: golang/gameoflife.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"math/rand"
	"os"
	"time"
)

var (
	rounds          int
	generationLimit int
	worldSize       int
	workers         int
	writeEveryGrid  bool
)

func init() {
	rand.Seed(time.Now().UnixNano())
	flag.IntVar(&rounds, "r", 10, "Simulation rounds")
	flag.IntVar(&generationLimit, "l", 2000, "Generation limit")
	flag.IntVar(&worldSize, "size", 20, "World size")
	flag.IntVar(&workers, "w", 4, "Number of workers")
	flag.BoolVar(&writeEveryGrid, "wa", false, "Write all generated grids to a file")
}

func main() {
	flag.Parse()

	if writeEveryGrid {
		os.Mkdir("grids", 0755)
	}

	bestSimulation := struct {
		score int
		sim   *simulation
	}{score: 0}

	type work struct {
		s        *simulation
		i, score int
	}

	simulationQueue := make(chan *work, 10)
	resultsQueue := make(chan *work, 10)
	// Start simulation runners
	for i := 0; i < workers; i++ {
		go func() {
			for w := range simulationQueue {
				w.score = w.s.run(generationLimit)
				resultsQueue <- w
			}
		}()
	}

	start := time.Now()
	// Start simulation generator
	go func() {
		for i := 1; i <= rounds; i++ {
			s := newSimulation(worldSize, worldSize, true)
			if writeEveryGrid {
				jsonData, _ := json.Marshal(s.currentState)
				ioutil.WriteFile(fmt.Sprintf("grids/simulation-%d.txt", i), jsonData, 0644)
			}
			simulationQueue <- &work{
				s: s,
				i: i,
			}
		}
		close(simulationQueue)
	}()

	// Collect and analyze simulation results
	for i := 1; i <= rounds; i++ {
		result := <-resultsQueue
		fmt.Printf("Simulation round %d\n", result.i)
		fmt.Printf("  Simulation score: %d\n", result.score)

		if result.score < generationLimit && result.score > bestSimulation.score {
			bestSimulation.score = result.score
			bestSimulation.sim = result.s
		}
	}

	close(resultsQueue)

	fmt.Printf("\nSimulation finished in %s\n", time.Since(start))

	fmt.Printf("\nBest score: %d\n", bestSimulation.score)
	jsonData, _ := json.Marshal(bestSimulation.sim.startGrid)
	ioutil.WriteFile("best.txt", jsonData, 0644)
}

type grid [][]int

func newGrid(width, height int) grid {
	g := make([][]int, height)
	for i := range g {
		g[i] = make([]int, width)
	}
	return g
}

func (g grid) randomize() {
	for i := range g {
		for j := range g[i] {
			if rand.Float32() < 0.2 {
				g[i][j] = 1
			}
		}
	}
}

func (g grid) copy() grid {
	newg := newGrid(len(g[0]), len(g))
	for i := range g {
		for j := range g[i] {
			newg[i][j] = g[i][j]
		}
	}
	return newg
}

type simulation struct {
	startGrid    grid
	nextState    grid
	currentState grid
	lastState    grid
}

func newSimulation(width, height int, randomize bool) *simulation {
	s := &simulation{
		nextState:    newGrid(width, height),
		currentState: newGrid(width, height),
		lastState:    newGrid(width, height),
	}

	if randomize {
		s.currentState.randomize()
	}
	return s
}

func (s *simulation) run(limit int) int {
	s.startGrid = s.currentState.copy()
	generation := 1
	stalled := false

	for !stalled {
		if limit > 0 && generation == limit {
			break
		}

		stalled = s.nextGeneration()
		s.rotateGrids()
		generation++
	}
	return generation
}

func (s *simulation) rotateGrids() {
	last := s.lastState
	s.lastState = s.currentState
	s.currentState = s.nextState
	s.nextState = last
}

func (s *simulation) nextGeneration() bool {
	stalled := true

	for i, row := range s.currentState {
		for j, cell := range row {
			neighbors := s.neighborCount(i, j)

			if neighbors < 2 {
				s.nextState[i][j] = 0
			} else if (neighbors == 2 || neighbors == 3) && cell == 1 {
				s.nextState[i][j] = 1
			} else if neighbors > 3 {
				s.nextState[i][j] = 0
			} else if neighbors == 3 && cell == 0 {
				s.nextState[i][j] = 1
			} else {
				s.nextState[i][j] = 0
			}

			if s.nextState[i][j] != s.lastState[i][j] {
				stalled = false
			}
		}
	}

	return stalled
}

func (s *simulation) neighborCount(row, column int) int {
	count := 0

	if row-1 >= 0 {
		if s.currentState[row-1][column] == 1 {
			count++
		}
	}

	if row+1 < len(s.currentState[0]) {
		if s.currentState[row+1][column] == 1 {
			count++
		}
	}

	if column-1 >= 0 {
		if s.currentState[row][column-1] == 1 {
			count++
		}
	}

	if column+1 < len(s.currentState) {
		if s.currentState[row][column+1] == 1 {
			count++
		}
	}

	if column-1 >= 0 && row-1 >= 0 {
		if s.currentState[row-1][column-1] == 1 {
			count++
		}
	}

	if column-1 >= 0 && row+1 < len(s.currentState[0]) {
		if s.currentState[row+1][column-1] == 1 {
			count++
		}
	}

	if column+1 < len(s.currentState) && row-1 >= 0 {
		if s.currentState[row-1][column+1] == 1 {
			count++
		}
	}

	if column+1 < len(s.currentState) && row+1 < len(s.currentState[0]) {
		if s.currentState[row+1][column+1] == 1 {
			count++
		}
	}

	return count
}
